### DISCUSSION TITLE
What is the difference between AsyncNotifier<Foo> and StateNotifier<AsyncValue<Foo>>

### DISCUSSION THREAD

**Discussion by meroo36** [rrousselGit/riverpod#3041] [Q&A] [answered]
Let me know if its not clear. Is there any difference?

**Comment by noriHanda** [comment]
AsyncNotifier has build method which allows you to watch another provider. This means that as the provider AsyncNotifier is watching changes, AsyncNotifier reacts to that change without recstructing itself.
StateNotifier on the other hand, gets fully reconstructed if it is watching another provider in its constructor. [edited]

That being said, I think it's clear that AsyncNotifier follows the definition of Riverpod better. i.e. "A Reactive Caching and Data-binding Framework"

**Reply by meroo36** [reply]
"StateNotifier on the other hand, only has contructor which only ignites once.", I am not sure if this statement is correct. I have used StateNotifiers many times like this:

```dart
final taskStateProvider =
    StateNotifierProvider<TaskStateNotifier, TaskState>((ref) {
  return TaskStateNotifier(ref);
});

class TaskStateNotifier extends StateNotifier<TaskState> {
  TaskStateNotifier(Ref ref)
      : _tasks = ref.watch(configProvider),
        _nftService = ref.watch(nftServiceProvider),
        _currentEvent = ref.watch(currentEventProvider),
        super(TaskState(<String, SingleTaskState>{})) {
    init();
  }
  ...
  ...
}
```
like this many times. By watching the providers, I can reconstruct the provider itself.

**Reply by noriHanda** [reply]
Thanks for pointing that out. I was wrong. Correct statement is that if you watch another provider from StateNotifier and that provider changes, the StateNotifier object itself is reconstructed while (Async)Notifier retains its state.
(I edited the original comment)

**Reply by meroo36** [reply]
I am still a bit confused about how AsyncNotifer doesn't reconstruct itself. If I watch anything inside AsyncNotifier the build method of the notifier will run which means it's being invalidated. It would be nice if you can give an example.

**Comment by AhmedLSayed9** [comment, marked as answer]
Check https://github.com/rrousselGit/riverpod/discussions/2831#discussioncomment-6781217

**Reply by meroo36** [reply]
thats pretty clear, thanks!
